hygroscopicity.F90

```diff
@@ -0,0 +1,90 @@
+! Include shortname defintions, so that the pre-existing F77 code does not have to be modified to
+! reference the CARMA structure.
+#include "carma_globaer.h"
+
+!! This routine calculates aerosol hygroscopicity
+!!
+!! @author  Pengfei Yu, Mike Mills
+!! @version Oct 2020
+subroutine hygroscopicity(carma, cstate, rc)
+  ! types
+  use carma_precision_mod
+  use carma_enums_mod
+  use carma_constants_mod
+  use carma_types_mod
+  use carmastate_mod
+  use carma_mod
+  use carmaelement_mod
+
+  implicit none
+
+  type(carma_type), intent(in)         :: carma   !! the carma object
+  type(carmastate_type), intent(inout) :: cstate  !! the carma state object
+  integer, intent(inout) :: rc      !! return code, negative indicates failure
+
+  ! Local declarations
+  integer         :: igroup         !! group index
+  integer         :: ibin           !! bin index
+  integer         :: iepart         !! element in group containing the particle concentration
+  integer         :: icore, i, z
+  real(kind=f)    :: coremass, shellmass
+  real(kind=f), parameter :: thresh  = 1e-10_f     ! relative threshold for core mass and kappa checking
+  real(kind=f), parameter :: thresh1 = 1._f+thresh
+  real(kind=f), parameter :: thresh0 = 0._f-thresh
+
+  1 format('hygroscopicity::ibin=',i4,',core mass=',e10.3,',shell mass=',e10.3,',hygroscopicity=',f9.4)
+
+  kappahygro(:NZ,:NBIN,:NGROUP) = -huge(1._f) ! default
+
+  ! loop through all bins, groups, and elements to calculate hygro for each group:
+  do igroup = 1,NGROUP
+    ! Only calculate hygro for groups that use it
+    if (irhswell(igroup) == I_PETTERS) then
+      iepart = ienconc(igroup)     ! element of particle number concentration
+      do ibin = 1, NBIN
+        do z = 1, NZ
+          kappahygro(z,ibin,igroup) = 0._f
+          if (pc(z, ibin, iepart).gt.0._f) then
+            ! Weight hygro by mass of each core
+            coremass = 0._f
+            do i = 1, ncore(igroup)
+              icore = icorelem(i, igroup)
+              coremass = coremass + pc(z, ibin, icore)
+              kappahygro(z,ibin,igroup) = kappahygro(z,ibin,igroup) + pc(z,ibin,icore) * kappaelem(icore)
+            end do ! i = 1, ncore(igroup)
+
+            if (coremass.gt.pc(z, ibin, iepart) * rmass(ibin, igroup) * thresh1) then
+              write(LUNOPRT,*) "hygro59: z,ibin,pc,rmass,coremass", &
+                z,ibin,pc(z, ibin, iepart),rmass(ibin, igroup),coremass
+              rc=RC_ERROR
+              return
+            end if
+
+            if (coremass.gt.pc(z, ibin, iepart) * rmass(ibin, igroup)) then
+              !Divide by total mass of all cores in the bin to normalize:
+              kappahygro(z,ibin,igroup) = kappahygro(z,ibin,igroup) / coremass
+            else
```

Comment:
Why the two cases for coremass? Should we just call the coremass check with the fix before entering this loop and then assume everything is set up correctly in this routine? Would make this code simpler and move the checking/repair to one place that will hopefully get removed in the future.

Reply:
I was not aware of the coremasscheck subroutine that Yunqian has written.

Reply:
I agree, we should just run the coremass check!

Reply:
deleted, done

Reply:
Still seeing the two cases. Perhaps I am not seeing the right versions of these files, but am looking at what GitHub is showing me.